### PR TITLE
Implement appointments and  addAppt command

### LINKED ISSUE #49: Implement "Add Appointment" feature
**Feature**: Add Appointment

**Purpose**: Allows doctors to schedule an appointment with a patient. 

**Command format**: `addAppt ApptName i/NRIC @d/DATE @t/TIME`

**Example commands**:
`addAppt Physio i/S1234567A @d/2024-09-20 @t/1000-1100`

**Acceptable values**:
- ApptName:
  - Must be Alphanumeric.
- NRIC: 
  - Must be 9 characters.
  - first and last characters being alphabets.
  - Follows the 9-character Singapore NRIC format for uniqueness.
  - Case-insensitive (e.g. ‘TXXXXXXXB’ is the same as ‘tXXXXXXXb’) for consistency.
- Date: 
  - Must be in the YYYY-MM-DD format (e.g., "2024-09-20").
  - Standard YYYY-MM-DD format for consistency and scheduling ease.
- Time: 
  - Must be in the HHMM format (24-hour clock) (e.g., "1000") 
  - must be in an interval (e.g. 1000-1100). 
  - Appointment time intervals cannot overlap with other existing appointment time intervals.
  - Standard HHMM format for consistency and scheduling ease.
  - Time format should be (start time - end time) where start time < end time.
The start time should not equal the end time. 

**Error message if invalid**:
- "Invalid date format. Use YYYY-MM-DD."
- "Invalid time format. Use HH:MM."
- "Patient does not exist.”
- “There is an existing appointment for this date and time”

**Outputs**:
- Success: "Appointment for John Doe scheduled on 2024-09-20 at 10:00."
- Failure: Error message related to the incorrect input format.
- Duplicate handling: The system will check if an appointment already exists for the given date and time interval, prompting a message like "Appointment already exists for this date and time."

### PR DESCRIPTION
**Pull request by yadobler**
fixes #49 

eg: `addAppt Physio i/S1234567A @d/2024-09-20 @t/1000-1100`

- Appointment class
- AddApptCommand class
- Appointment saving in json
- Appointment UI like tags
- Each patient cannot have clashing Appointments

TODO: 
- Tests for AddApptCommand and parser
- Update UG

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 1]
```diff
@@ -56,8 +57,9 @@ public AddCommand parse(String args) throws ParseException {
         Email email = ParserUtil.parseEmail(argMultimap.getValue(PREFIX_EMAIL).get());
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
         Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
+        Set<Appointment> appointmentList = ParserUtil.parseAppointments(argMultimap.getAllValues(PREFIX_TAG));
 
-        Person person = new Person(name, phone, email, nric, address, dob, gender, tagList);
+        Person person = new Person(name, phone, email, nric, address, dob, gender, tagList, appointmentList);
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:62]: For `appointmentList`,  are you planning to make it such that the user can choose to also input the patient's appt using the existing `t/` tag prefix when adding the patient and it will get parsed by `parseappointment` and only added if it is a valid appointment?

If not, should we set  `appointmentList` to be an empty set first? So that the `Person` object will be initialise with no appts.

File: src/main/java/seedu/address/model/appointment/Appointment.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * Returns true if a given string is a valid appointment name.
+     */
+    public static boolean isValidAppointmentName(String test) {
+        return test.matches(NAME_VALIDATION_REGEX);
+    }
+
+    /**
+     *  checks if a given string is a valid appointment time period.
+     */
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/model/appointment/Appointment.java:109]: Should caps the first letter for javadocs and remove the whitespace

File: src/main/java/seedu/address/model/appointment/Appointment.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                      MESSAGE_CONSTRAINTS_APPT_TIME_PERIOD_INVALID_ORDER);
+    }
+
+    private static Boolean isValidAppointmentTimePeriodFormat(String test) {
+        return test.matches(TIME_VALIDATION_REGEX);
+    }
+
+    private static Boolean isValidAppointmentTimePeriodOrder(String test) {
+        return Integer.parseInt(test.substring(0, 4)) < Integer.parseInt(test.substring(5, 9));
+    }
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/model/appointment/Appointment.java:123]: Functions following naming conventions 👍🏻

File: src/test/java/seedu/address/model/appointment/AppointmentTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        Assertions.assertTrue(0 < a.compareTo(sameDateBefore));
+        Assertions.assertEquals(0, a.compareTo(sameDateTime));
+        Assertions.assertTrue(0 > a.compareTo(sameDateAfter));
+        Assertions.assertTrue(0 < a.compareTo(earlyDateBefore));
+        Assertions.assertTrue(0 < a.compareTo(earlyDateSame));
+        Assertions.assertTrue(0 < a.compareTo(earlyDateAfter));
+        Assertions.assertTrue(0 > a.compareTo(laterDateBefore));
+        Assertions.assertTrue(0 > a.compareTo(laterDateSame));
+        Assertions.assertTrue(0 > a.compareTo(laterDateAfter));
+    }
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/model/appointment/AppointmentTest.java:97]: Test cases are comprehensive and well designed 👍🏻

File: src/main/java/seedu/address/logic/commands/AddApptCommand.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        Person personToEdit = model.fetchPersonIfPresent(predicate)
+                                   .orElseThrow(() -> new CommandException(Messages.MESSAGE_PERSON_NRIC_NOT_FOUND));
+        Person editedPerson = createEditedPerson(personToEdit, newApptName, newApptDate, newApptTime);
+        model.setPerson(personToEdit, editedPerson);
+        logger.info("AddAppt command has been exceuted successfully on a person with NRIC: " + predicate);
+        model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
+        return new CommandResult(String.format(MESSAGE_SUCCESS_2S, predicate, String.format("%s [ %s @ %s ]",
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/commands/AddApptCommand.java:86]: Good use of logging

**Review by yong-zhuo — COMMENTED** [review]
LGTM overall 👍🏻, just need to clarify the changes made in `AddCommandParser`

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java (hunk 1 above)
Comment (yadobler) [reply src/main/java/seedu/address/logic/parser/AddCommandParser.java:62]: Hm I think it's a good idea, but I think the problem is each appt takes in 3 arguments, so I think it might be too hassle for when adding new patient. 

For now I will stick to making it default to empty list first. Maybe it be another feature.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 6]
```diff
@@ -169,6 +171,29 @@ public static Tag parseTag(String tag) throws ParseException {
...
 
+    /**
+     * Parses a {@code String appointment} into a {@code Appointment}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code appointment} is invalid.
+     */
+    public static Appointment parseAppointment(String appointment) throws ParseException {
+        requireNonNull(appointment);
+        String[] trimmedAppointment = appointment.trim().split(":");
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:182]: Could perhaps be renamed to trimmedAppointments

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 7]
```diff
@@ -198,4 +223,20 @@ public static Priority parsePriority(String priorityStr) throws ParseException {
         }
         return new Priority(trimmedPriority);
     }
+
+    /**
+     * Parses {@code Collection<String> appointments} into a {@code Set<Appointment>}.
+     */
+    public static Set<Appointment> parseAppointments(Collection<String> appointments) throws ParseException {
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:230]: Just a nit: Could have a better naming for parseAppointments as parseAppointment exists, this might cause some bugs due to naming. Maybe parseMultipleAppointment?

**Review by hweinian — COMMENTED** [review]
A few nits here and there in terms of naming, code duplication. Other than that LGTM

File: src/test/java/seedu/address/model/appointment/AppointmentTest.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        Assertions.assertFalse(a.isClashing("2024-10-25", "0900-0935")); // "sameDateBefore"
+        Assertions.assertTrue(a.isClashing("2024-10-25", "0900-1235")); // "sameDateBefore"
+        Assertions.assertTrue(a.isClashing("2024-10-25", "0845-1300")); // "sameDateTimeEnclose"
+        Assertions.assertTrue(a.isClashing("2024-10-25", "1000-1235")); // "sameDateTimeInside"
+        Assertions.assertTrue(a.isClashing("2024-10-25", "1045-1200")); // "sameDateTime"
+        Assertions.assertTrue(a.isClashing("2024-10-25", "1100-1235")); // "sameDateAfter"
+        Assertions.assertFalse(a.isClashing("2024-10-25", "1235-1345")); // "sameDateAfter"
+
+        Assertions.assertFalse(a.isClashing("2024-09-25", "0900-0935")); // "earlierDateBefore"
+        Assertions.assertFalse(a.isClashing("2024-09-25", "0900-1235")); // "earlierDateBefore"
```
[the remark this answers is not in the document]
Comment (yadobler) [reply src/test/java/seedu/address/model/appointment/AppointmentTest.java:150]: I think I was facing an edge case where if the appointments ended or began at the same time, the code will return false (ie not clashing). I will review and see how I went through it, and also comment the process.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 7 above)
Comment (yadobler) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:230]: Right! I think this is mainly for manual parsing / json text parsing, so I'll see how I can rename it (since appt takes multiple strings to construct)

File: src/main/java/seedu/address/model/appointment/Appointment.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        if (timePeriod.equals(appointmentTimePeriod)) {
+            return true;
+        }
+
+        boolean isTimePeriodStartingDuringAppt = startTime.isAfter(appointmentStartTime)
+                                                 && startTime.isBefore(appointmentEndTime);
+        boolean isTimePeriodEndingDuringAppt = endTime.isAfter(appointmentStartTime)
+                                               && endTime.isBefore(appointmentEndTime);
+        boolean isApptWithinTimePeriod = startTime.isBefore(appointmentStartTime)
+                                         && endTime.isAfter(appointmentEndTime);
```
[the remark this answers is not in the document]
Comment (yadobler) [reply src/main/java/seedu/address/model/appointment/Appointment.java]: Ok I have simplified it. #cs1231s
